File: frontend/src/services/api/calculationAssumptions.ts
/**
 * Calculation Assumption Registry API client.
 *
 * Phase 2 endpoints (catalog, list/get/effective/history/dependencies, write
 * lifecycle) plus Phase 5 variance reporting.
 */

import api from './client'

// ----------------------------------------------------------- shared shapes

export type AssumptionStatus = 'proposed' | 'active' | 'retired'

export interface AssumptionResponse {
  assumption_id: number
  client_id: string
  assumption_name: string
  value: unknown
  rationale: string | null
  effective_date: string | null
  expiration_date: string | null
  status: AssumptionStatus
  proposed_by: string
  proposed_at: string
  approved_by: string | null
  approved_at: string | null
  retired_by: string | null
  retired_at: string | null
  is_active: boolean
  created_at: string
  updated_at: string | null
}

export interface CatalogEntry {
  name: string
  description: string
  allowed_values: unknown[] | null
}

// ------------------------------------------------------------- Phase 5

export interface VarianceRow {
  assumption_id: number
  client_id: string
  assumption_name: string
  description: string | null
  value: unknown
  default_value: unknown
  deviates_from_default: boolean
  deviation_magnitude: number
  approved_by: string | null
  approved_at: string | null
  days_since_review: number | null
  is_stale: boolean
  rationale: string | null
}

// Paths are relative to the `api` client's baseURL (/api/v1, rewritten to
// /api server-side) — a literal leading `/api/` here double-prefixes the
// request (e.g. /api/v1/api/assumptions/variance), which 404s.
export const getCatalog = () => api.get<CatalogEntry[]>('/assumptions/catalog')

export const getVarianceReport = (staleAfterDays = 365) =>
  api.get<VarianceRow[]>('/assumptions/variance', {
    params: { stale_after_days: staleAfterDays },
  })

export const listAssumptions = (params?: {
  client_id?: string
  assumption_name?: string
  status?: AssumptionStatus
  include_inactive?: boolean
}) => api.get<AssumptionResponse[]>('/assumptions', { params })

// ------------------------------------------------- the write lifecycle
//
// None of the below had a client function, so the entire propose -> approve
// -> retire workflow -- and the append-only change log behind it -- was
// unreachable from the product. The site-adjusted half of the dual view is
// driven by approved assumptions, so the only way to get one was direct DB
// access.
//
// AUTH IS NARROWER THAN THE ROUTE SUGGESTS. Each write passes a FastAPI
// dependency AND a second, tighter check inside AssumptionService, and only
// the second one is the real answer:
//
//   propose   route: supervisory   service: admin | poweruser
//   edit      route: supervisory   service: the original proposer, or admin
//   approve   route: planner       service: ADMIN ONLY
//   retire    route: planner       service: ADMIN ONLY
//
// Gating a button on the route's tier would therefore offer approve to a
// poweruser and 403 them. Gate on admin.

export interface AssumptionChangeRow {
  change_id: number
  assumption_id: number
  changed_by: string
  changed_at: string
  previous_value: unknown | null
  new_value: unknown | null
  previous_status: string | null
  new_status: string | null
  change_reason: string | null
  trigger_source: string | null
}

export interface ProposalPayload {
  client_id: string
  assumption_name: string
  value: unknown
  rationale?: string | null
  effective_date?: string | null
  expiration_date?: string | null
}

export interface ProposalPatch {
  value?: unknown
  rationale?: string | null
  effective_date?: string | null
  expiration_date?: string | null
  change_reason?: string | null
}

export const getAssumption = (assumptionId: number) =>
  api.get<AssumptionResponse>(`/assumptions/${assumptionId}`)

/** The append-only governance trail: who changed what, when, and why. */
export const getAssumptionHistory = (assumptionId: number) =>
  api.get<AssumptionChangeRow[]>(`/assumptions/${assumptionId}/history`)

export const getEffectiveSet = (clientId: string, asOf?: string) =>
  api.get('/assumptions/effective', {
    params: asOf ? { client_id: clientId, as_of: asOf } : { client_id: clientId },
  })

export const proposeAssumption = (data: ProposalPayload) =>
  api.post<AssumptionResponse>('/assumptions', data)

/** Only while PROPOSED -- the API answers 409 once a record is active. */
export const updateProposal = (assumptionId: number, data: ProposalPatch) =>
  api.patch<AssumptionResponse>(`/assumptions/${assumptionId}`, data)

export const approveAssumption = (assumptionId: number, changeReason?: string | null) =>
  api.post<AssumptionResponse>(`/assumptions/${assumptionId}/approve`, {
    change_reason: changeReason ?? null,
  })

export const retireAssumption = (assumptionId: number, changeReason?: string | null) =>
  api.post<AssumptionResponse>(`/assumptions/${assumptionId}/retire`, {
    change_reason: changeReason ?? null,
  })

File: backend/routes/calculation_assumptions.py
"""
Calculation assumption registry — Phase 2 dual-view architecture.

REST API for proposing, approving, retiring, and querying site-level
calculation assumptions.

Role gates (option B ratified):
  - admin     → approve, retire (status transitions)
  - poweruser → propose, edit own proposal
  - others    → read only
"""

from datetime import datetime
from typing import Any, List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from backend.auth.jwt import get_current_user
from backend.database import get_db
from backend.orm.calculation_assumption import AssumptionStatus
from backend.orm.user import User
from backend.schemas.calculation_assumption import (
    AssumptionApproveRequest,
    AssumptionChangeResponse,
    AssumptionProposalCreate,
    AssumptionProposalUpdate,
    AssumptionResponse,
    AssumptionRetireRequest,
    AssumptionStatusEnum,
    CatalogEntry,
    EffectiveAssumptionSet,
    MetricAssumptionDependencyResponse,
    VarianceRow,
)
from backend.services.assumption_service import AssumptionService
from backend.utils.logging_utils import get_module_logger

logger = get_module_logger(__name__)

router = APIRouter(prefix="/api/assumptions", tags=["Calculation Assumptions"])


# --------------------------------------------------------------------- helpers


def _to_response(record) -> AssumptionResponse:
    import json

    return AssumptionResponse(
        assumption_id=record.assumption_id,
        client_id=record.client_id,
        assumption_name=record.assumption_name,
        value=json.loads(record.value_json),
        rationale=record.rationale,
        effective_date=record.effective_date,
        expiration_date=record.expiration_date,
        status=AssumptionStatusEnum(record.status.value),
        proposed_by=record.proposed_by,
        proposed_at=record.proposed_at,
        approved_by=record.approved_by,
        approved_at=record.approved_at,
        retired_by=record.retired_by,
        retired_at=record.retired_at,
        is_active=bool(record.is_active),
        created_at=record.created_at,
        updated_at=record.updated_at,
    )


def _to_change_response(change) -> AssumptionChangeResponse:
    import json

    return AssumptionChangeResponse(
        change_id=change.change_id,
        assumption_id=change.assumption_id,
        changed_by=change.changed_by,
        changed_at=change.changed_at,
        previous_value=json.loads(change.previous_value_json) if change.previous_value_json else None,
        new_value=json.loads(change.new_value_json) if change.new_value_json else None,
        previous_status=change.previous_status,
        new_status=change.new_status,
        change_reason=change.change_reason,
        trigger_source=change.trigger_source,
    )


# ------------------------------------------------------------------------ reads


@router.get("/catalog", response_model=List[CatalogEntry])
def get_catalog(current_user: User = Depends(get_current_user)) -> List[CatalogEntry]:
    """Static v1 catalog of valid assumption names + their permitted values."""

    catalog = AssumptionService.get_catalog()
    return [
        CatalogEntry(
            name=name,
            description=entry["description"],
            allowed_values=entry.get("allowed_values"),
        )
        for name, entry in catalog.items()
    ]


@router.get("/dependencies", response_model=List[MetricAssumptionDependencyResponse])
def list_dependencies(
    metric_name: Optional[str] = None,
    assumption_name: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> List[MetricAssumptionDependencyResponse]:
    """Map of which metrics consume which assumptions (engineering-curated)."""

    rows = AssumptionService(db, current_user).list_dependencies(
        metric_name=metric_name, assumption_name=assumption_name
    )
    return [MetricAssumptionDependencyResponse.model_validate(r) for r in rows]


@router.get("/variance", response_model=List[VarianceRow])
def get_variance_report(
    stale_after_days: int = Query(365, ge=1, le=3650),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> List[VarianceRow]:
    """Phase 5 — adjustment variance report.

    Returns one row per ACTIVE assumption (across all sites for admin/poweruser;
    tenant-scoped otherwise). Each row carries the catalog default + a deviation
    magnitude suitable for sortable-table UIs, plus an `is_stale` flag for
    assumptions not reviewed in `stale_after_days` (default 365).
    """

    rows = AssumptionService(db, current_user).get_variance_report(stale_after_days=stale_after_days)
    return [VarianceRow(**row) for row in rows]


@router.get("/effective", response_model=EffectiveAssumptionSet)
def get_effective_set(
    client_id: str = Query(..., description="Client to fetch the effective assumption set for"),
    as_of: Optional[datetime] = Query(None, description="ISO-8601 timestamp; defaults to now (UTC)"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> EffectiveAssumptionSet:
    """Return the active assumption value per name for a client at a point in time."""

    import json
    from datetime import timezone

    effective = AssumptionService(db, current_user).get_effective_set(client_id=client_id, as_of=as_of)
    as_of_value = as_of or datetime.now(tz=timezone.utc)

    assumptions: dict[str, Any] = {}
    metadata: dict[str, AssumptionResponse] = {}
    for name, record in effective.items():
        assumptions[name] = json.loads(record.value_json)
        metadata[name] = _to_response(record)

    return EffectiveAssumptionSet(
        client_id=client_id,
        as_of=as_of_value,
        assumptions=assumptions,
        metadata=metadata,
    )


@router.get("", response_model=List[AssumptionResponse])
def list_assumptions(
    client_id: Optional[str] = None,
    assumption_name: Optional[str] = None,
    status_filter: Optional[AssumptionStatusEnum] = Query(None, alias="status"),
    include_inactive: bool = False,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> List[AssumptionResponse]:
    """List assumption records, with tenant-isolation enforced for non-admin/poweruser users."""

    status_enum = AssumptionStatus(status_filter.value) if status_filter else None
    rows = AssumptionService(db, current_user).list_assumptions(
        client_id=client_id,
        assumption_name=assumption_name,
        status_filter=status_enum,
        include_inactive=include_inactive,
    )
    return [_to_response(r) for r in rows]


@router.get("/{assumption_id}", response_model=AssumptionResponse)
def get_assumption(
    assumption_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> AssumptionResponse:
    record = AssumptionService(db, current_user).get_assumption(assumption_id)
    return _to_response(record)


@router.get("/{assumption_id}/history", response_model=List[AssumptionChangeResponse])
def get_history(
    assumption_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> List[AssumptionChangeResponse]:
    """Append-only change log for one assumption record (newest first)."""

    rows = AssumptionService(db, current_user).get_history(assumption_id)
    return [_to_change_response(r) for r in rows]


# ----------------------------------------------------------------------- writes


@router.post("", response_model=AssumptionResponse, status_code=status.HTTP_201_CREATED)
def propose_assumption(
    body: AssumptionProposalCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> AssumptionResponse:
    """Submit a new assumption proposal. Requires `poweruser` or `admin` role."""

    record = AssumptionService(db, current_user).propose(
        client_id=body.client_id,
        assumption_name=body.assumption_name,
        value=body.value,
        rationale=body.rationale,
        effective_date=body.effective_date,
        expiration_date=body.expiration_date,
    )
    return _to_response(record)


@router.patch("/{assumption_id}", response_model=AssumptionResponse)
def update_proposal(
    assumption_id: int,
    body: AssumptionProposalUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> AssumptionResponse:
    """Edit a PROPOSED record. Allowed for the original proposer or any admin."""

    fields_set = body.model_fields_set
    record = AssumptionService(db, current_user).update_proposal(
        assumption_id=assumption_id,
        value=body.value if "value" in fields_set else None,
        rationale=body.rationale if "rationale" in fields_set else None,
        effective_date=body.effective_date if "effective_date" in fields_set else None,
        expiration_date=body.expiration_date if "expiration_date" in fields_set else None,
        change_reason=body.change_reason,
        effective_date_provided="effective_date" in fields_set,
        expiration_date_provided="expiration_date" in fields_set,
    )
    return _to_response(record)


@router.post("/{assumption_id}/approve", response_model=AssumptionResponse)
def approve_assumption(
    assumption_id: int,
    body: AssumptionApproveRequest = AssumptionApproveRequest(),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> AssumptionResponse:
    """Transition PROPOSED → ACTIVE. Admin only. Auto-retires overlapping ACTIVE records."""

    record = AssumptionService(db, current_user).approve(assumption_id=assumption_id, change_reason=body.change_reason)
    return _to_response(record)


@router.post("/{assumption_id}/retire", response_model=AssumptionResponse)
def retire_assumption(
    assumption_id: int,
    body: AssumptionRetireRequest = AssumptionRetireRequest(),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> AssumptionResponse:
    """Transition ACTIVE → RETIRED. Admin only."""

    record = AssumptionService(db, current_user).retire(assumption_id=assumption_id, change_reason=body.change_reason)
    return _to_response(record)
